core/chains/evm/txmgr/finalizer.go

```diff
@@ -0,0 +1,274 @@
+package txmgr
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/rpc"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+	"github.com/smartcontractkit/chainlink-common/pkg/utils/mailbox"
+
+	evmtypes "github.com/smartcontractkit/chainlink/v2/core/chains/evm/types"
+)
+
+var _ Finalizer = (*evmFinalizer)(nil)
+
+// processHeadTimeout represents a sanity limit on how long ProcessHead
+// should take to complete
+const processHeadTimeout = 10 * time.Minute
+
+type finalizerTxStore interface {
+	FindConfirmedTxesAwaitingFinalization(ctx context.Context, chainID *big.Int) ([]*Tx, error)
+	UpdateTxesFinalized(ctx context.Context, txs []int64, chainId *big.Int) error
+}
+
+type finalizerChainClient interface {
+	BatchCallContext(ctx context.Context, elems []rpc.BatchElem) error
+}
+
+// Finalizer handles processing new finalized blocks and marking transactions as finalized accordingly in the TXM DB
+type evmFinalizer struct {
+	services.StateMachine
+	lggr         logger.SugaredLogger
+	chainId      *big.Int
+	rpcBatchSize int
+	txStore      finalizerTxStore
+	client       finalizerChainClient
+	mb           *mailbox.Mailbox[*evmtypes.Head]
+	stopCh       services.StopChan
+	wg           sync.WaitGroup
+}
+
+func NewEvmFinalizer(
+	lggr logger.Logger,
+	chainId *big.Int,
+	rpcBatchSize uint32,
+	txStore finalizerTxStore,
+	client finalizerChainClient,
+) *evmFinalizer {
+	lggr = logger.Named(lggr, "Finalizer")
+	return &evmFinalizer{
+		lggr:         logger.Sugared(lggr),
+		chainId:      chainId,
+		rpcBatchSize: int(rpcBatchSize),
+		txStore:      txStore,
+		client:       client,
+		mb:           mailbox.NewSingle[*evmtypes.Head](),
+	}
+}
+
+// Start is a comment to appease the linter
+func (f *evmFinalizer) Start(ctx context.Context) error {
+	return f.StartOnce("Finalizer", func() error {
+		return f.startInternal(ctx)
+	})
+}
+
+func (f *evmFinalizer) startInternal(_ context.Context) error {
```

Comment:
You don't need `startInternal` scheme, this is only for services that require to refetch the enabled addresses via `EnabledAddressesForChain` . You should follow the same pattern as `reaper.go`.

Reply:
I ran into some problems with health checks when I followed the Reaper pattern. I removed `startinternal` but still kept the `StartOnce` pattern which has some health status changes built in.